staging/src/k8s.io/kube-apiextensions-server/pkg/controller/finalizer/crd_finalizer.go

```
@@ -0,0 +1,297 @@
+		apiextensions.SetCRDCondition(crd, apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.Terminating,
+			Status:  apiextensions.ConditionTrue,
+			Reason:  "InstanceDeletionStuck",
+			Message: fmt.Sprintf("cannot proceed with deletion because of %v condition", apiextensions.NameConflict),
+		})
+		crd, err = c.crdClient.CustomResourceDefinitions().UpdateStatus(crd)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("cannot proceed with deletion because of %v condition", apiextensions.NameConflict)
+	}
+
+	// Now we can start deleting items.  We should use the REST API to ensure that all normal admission runs.
+	// Since we control the endpoints, we know that delete collection works.
+	crClient, err := c.clientPool.ClientForGroupVersionResource(schema.GroupVersionResource{Group: crd.Spec.Group, Version: crd.Spec.Version, Resource: crd.Status.AcceptedNames.Plural})
+	if err != nil {
+		return err
+	}
+	crAPIResource := &metav1.APIResource{
+		Name:         crd.Status.AcceptedNames.Plural,
+		SingularName: crd.Status.AcceptedNames.Singular,
+		Namespaced:   crd.Spec.Scope == apiextensions.NamespaceScoped,
+		Kind:         crd.Status.AcceptedNames.Kind,
+		Verbs:        metav1.Verbs([]string{"deletecollection", "list"}),
+		ShortNames:   crd.Status.AcceptedNames.ShortNames,
+	}
+	crResourceClient := crClient.Resource(crAPIResource, "" /* namespace all */)
+	allResources, err := crResourceClient.List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	deletedNamespaces := sets.String{}
+	for _, item := range allResources.(*unstructured.UnstructuredList).Items {
+		metadata, err := meta.Accessor(&item)
+		if err != nil {
+			utilruntime.HandleError(err)
+			continue
+		}
+		if deletedNamespaces.Has(metadata.GetNamespace()) {
+			continue
+		}
+		if err := crClient.Resource(crAPIResource, metadata.GetNamespace()).DeleteCollection(nil, metav1.ListOptions{}); err != nil {
+			utilruntime.HandleError(err)
+			continue
```

If one of the deletes fails, shouldn't we return an error (skip the wait step below) to trigger a requeue? Currently I guess it will retry if there are multiple items in the same namespace, but not if there is only one. And as it is, if there are many items in the same namespace we may end up hammering `DeleteCollection()` on error without backoff.